presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java

```diff
@@ -261,6 +261,59 @@ public RowExpression pushNegationToLeaves(RowExpression expression)
         return expression.accept(new PushNegationVisitor(), null);
     }
 
+    /**
+     * Given a logical expression, the goal is to convert to conjuctive normal form (CNF).
+     * This requires making a call to `pushNegationToLeaves`. If there are more than two clauses,
+     * the tree is constructed as a right-deep tree (i.e. X AND (Y AND (Z AND ...)))).
```

Comment:
Just curious why choosing right-deep?

Reply:
Actually, being right-deep is incorrect. I'll update the comment. There's no guarantee as to the balance of tree. Every time we need to "push down" a termJoiner (swap AND and OR) then we'll balance the tree by collecting the conjuncts/disjuncts and constructing a new Expression, else we just append left and right expressions in which case if 1 side is always left or right heavy,  this will propagate.

I could make it always balance the tree, which wouldn't impact the worst case performance, every time we visit a conjunct/disjunct.

Reply:
I was wondering why did we choose right-deep rather than a balanced tree or left-deep. 😂 I personally would go in that order 😛

Reply:
Do you suggest I enforce the balanced tree always or leave it as is?

Reply:
I see. It's whatever the tree shape `combinePredicates` returns :).